### ISSUE TITLE
[FRONTEND] Support Tensorflow

### ISSUE THREAD

**Issue by xqdan** [dmlc/nnvm#234]

**Comment by tqchen** [comment]
There is no on going effort in the community atm, but contribution is welcomed.

**Comment by bonotake** [comment]
Now our group is trying to import TF protobuf files via ONNX. If it succeeds well, we may PR then.

**Comment by bonotake** [comment]
Of course if anyone tries to import directly from TF, it’s very welcome for us, too.

**Comment by xqdan** [comment]
@bonotake that's great! do you have a ballpark time for this PR?

**Comment by bonotake** [comment]
@xqdan actually no, we're in just a preliminary trial phase. We'll check the feasibility for our own purpose, then determine whether to publish it as it is feasible for general purpose.

**Comment by masahi** [comment]
Looks like there is a tensorflow backend support for onnx in development. 
See [here](https://github.com/onnx/onnx-tensorflow) and onnx/onnx#185.

**Comment by bonotake** [comment]
@masahi I know it but it's the opposite direction: ONNX -> TensorFlow.

**Comment by xqdan** [comment]
Google today released a tool that converts AI models produced for mobile devices using its TensorFlow Lite tool into Apple’s Core ML .mlmodel file format for use with iOS devices.
https://venturebeat.com/2017/12/05/google-brings-core-ml-support-to-tensorflow-lite/

**Comment by guopinglong** [comment]
Transforming TF Graphdef into a NNVM graph? Sounds interesting.

**Comment by PariksheetPinjari909** [comment]
Some update from my side on Tensorflow.

I could transform Graphdef to NNVM and it works fine for Inception V1 & Inception V3.
It has multiple tasks as described below.

- [x] DecodeJpeg operator support through extern interface needed for V1 (NNVM & TVM).
- [x] Passing scheme issue as described in https://github.com/dmlc/nnvm/issues/261 and https://github.com/dmlc/nnvm/issues/371
- [x] Tensorflow frontend.
- [x] Tutorial examples for tensorflow (V1 & V3).
- [ ] Matmul issue as described in https://github.com/dmlc/nnvm/issues/373 - Need help
- [x] some bugfix to sqeeze operator
- [ ] Test cases for Tensorflow.

Working on cleanup & the completed soon.

Siva

**Comment by fumihwh** [comment]
FYI, I am working at onnx-tensorflow. onnx-tensorflow can convert tensorflow graph to onnx. 
Of course support tensorflow directly sounds cool.

**Comment by FrozenGene** [comment]
@PariksheetPinjari909 Are you going to pull the request of tf-frontend? Because I will start to work on the tf-frontend this week. I don't want to do the duplicate work if we have existing one. Or I can do the work based on yours if you would pull the request.

**Comment by PariksheetPinjari909** [comment]
@srkreddy1238 Can you update on tensorflow frontend progress ?

**Comment by srkreddy1238** [comment]
PR for image ops
https://github.com/dmlc/tvm/pull/1141
https://github.com/dmlc/nnvm/pull/464

**Comment by srkreddy1238** [comment]
@tqchen  : I checked the keras frontend about dynamic padding & my suspect was right :)

Keras is not actually a dynamic padding as the input width and height is available at frontend import from keras_layer info- https://github.com/dmlc/nnvm/blob/8c0a103aaeb7daf9a986732340e3482cd8b1a2ed/python/nnvm/frontend/keras.py#L155

Tensorflow is different and we don't have layer input dimensions at model importing. Actuall input shapes will be available only while nnvm build.

Pls advice.


Siva

**Comment by tqchen** [comment]
@srkreddy1238 One way to do so is that we can provide the input shape during frontend.from_tensorflow

**Comment by fumihwh** [comment]
@srkreddy1238 You can get all `_output_shapes` from nodes and pass them to your handlers.

**Comment by srkreddy1238** [comment]
@tqchen 

Handling of SAME in tensorflow depends on the input dimensions of the conv (or pooling ..etc.) node.

The frontend and nnvm.build are two different modules now in nnvm architecture.
Infer shape happens only in nnvm build process now when the "graph input dimensions" are passed.

I suggest with a dedicated pad operator we already have, the pad operator schema can be changed to handle build time calculation of padding via padding_scheme. This avoids schemantic change of many operators. 

_output_shapes ? : Can you share more details ? 
Unless importing graph layers has static defines for input and out shapes of layers, it's not possible to get shapes without calling InterShape which is now happening while nnvm build only.

Pls advice.

**Comment by srkreddy1238** [comment]
@Huyuwei : For tensor flow pad calculation happens after InferShape of previous nodes is completed not while importing the graph.

**Comment by fumihwh** [comment]
@srkreddy1238 
```
pseudo code:

...
for node in graph_def.node:
	output_shapes[node.name] = node.attr["_output_shapes"]
	...


def _convert_convolution(insym, ..., output_shapes):
	input_shape = output_shapes[input_name]

```

**Comment by srkreddy1238** [comment]
@fumihwh 

Thanks. I will check on this.

**Comment by srkreddy1238** [comment]
@fumihwh 

I checked add_shapes option, this option need to be added before generating the graph.pb.

How about the case with Google published models like Inception V1, V3 ...etc which don't have _output_shapes attributes.

The sample graph definition before nnvm front end call below.

`def _ProcessGraphDefParam(graph_def):
  if not isinstance(graph_def, graph_pb2.GraphDef):
    try:
      old_graph_def = graph_def
      graph_def = graph_pb2.GraphDef()
      graph_def.MergeFrom(old_graph_def)
    except TypeError:
      raise TypeError('graph_def must be a GraphDef proto.')
  return graph_def

with tf.gfile.FastGFile(os.path.join(
        "./", 'classify_image_graph_def.pb'), 'rb') as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())
    graph = tf.import_graph_def(graph_def, name='')
    graph_def = _ProcessGraphDefParam(graph_def)

sym, params = nnvm.frontend.from_tensorflow(graph_def)`


Do you suggest to reload the graph over a tensorflow session then freeze with add_shapes=true before sending it to frontend ?

**Comment by srkreddy1238** [comment]
@tqchen what do you suggest ?

**Comment by srkreddy1238** [comment]
@tqchen 

You mean developer pass shapes as argument to from_tensorflow call ?

**Comment by srkreddy1238** [comment]
Ok

**Comment by srkreddy1238** [comment]
By _output_shapes compiled into the tensorflow export here is the PR for Tensorflow frontend (with Inception V1).

https://github.com/dmlc/nnvm/pull/472

**Comment by srkreddy1238** [comment]
Updated task list

- [x] Tensorflow frontend (V1)
- [x] More Models (V3 & others)
- [x] Tutorials.
- [ ] Test cases.

**Comment by tqchen** [comment]
move to https://github.com/dmlc/tvm/issues/1186
